### PR TITLE
add śiva post.

### PR DESCRIPTION
**Pull request by smola**

### CODE REVIEW

**Comment by Serabe** [comment]
Awesome post :)

File: content/post/siva.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+Here is a table with a comparison of different archive formats with respect our
+requirements.
+ 
+| Format | No compression | Seekable | Indexed | Cheap Add | Concatenable |
+| ------ | -------------- | -------- | ------- | --------- | ------------ |
+| [zip](https://en.wikipedia.org/wiki/Zip_(file_format)) | ✔ | ✔[^zipseek] | ✔ | ✔ | ✘ |
+| [tar](https://www.freebsd.org/cgi/man.cgi?query=tar&sektion=5) | ✔ | ✔ | ✘ | ✔ | ✔[^tarconcat] | 
+| [cpio](https://www.freebsd.org/cgi/man.cgi?query=cpio&sektion=5) | ✔ | ✔ | ✘ | ✔ | ✘[^cpioconcat] |
+| [7z](http://www.7-zip.org/7z.html) | ✔ | ✔[^7zseek] | ✔ | ✔ | ✘ |
```
Comment (bzz) [inline content/post/siva.md]: indeed, it is :+1:

File: content/post/siva.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+   implementation is needed for every language that we use.
+2. It is not space efficient: metadata is duplicated both in the index and headers,
+   plus the useless padding. This is not a big issue for us, but if we are going
+   to write our own tar implementations, we would rather avoid cruft and legacy.
+
+## zip
+
+Zip seems to come pretty close to our requirements. Let's see its structure:
+
+![tar file structure](/post/siva/zip.png)
```
Comment (bzz) [inline content/post/siva.md]: s/tar /zip /

File: content/post/siva.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+ and its index file.
+ * `HEAD` file with the HEAD reference.
+ * (optional) `refs` directory containing [references](https://git-scm.com/book/en/v2/Git-Internals-Git-References).
+ * (optional) `packed-refs` file with [packed references](https://git-scm.com/docs/git-pack-refs).
+
+For each repository, we store, at least, 5 files (config, 1 packfile,
+1 packfile index, HEAD reference and a master reference). This grows, at least,
+by two (1 incremental packfile and its index) every time with update a repository.
+
+For a fresh fetch of our current ~16M repository, that adds up to 80M files and
```
Comment (smola) [inline content/post/siva.md]: Done.

File: content/post/siva.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+32M more for each update. That is without considering the high monthly growth of
+repository number itself.
+
```
Comment (smola) [inline content/post/siva.md]: :+1:

File: content/post/siva.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+We store all of this in a distributed file system. Currently Google Cloud
```
Comment (smola) [inline content/post/siva.md]: replaced with "everything" instead

File: content/post/siva.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+repository to a computing node.
+
+At this point, it was clear to us that we need to archive repositories. That way,
+when processing a repository, we would need to fetch a single file from the
+distributed file system.
+
+## Our Ideal Archive Format
+
+So we started evaluating different archiving formats to use. That should be easy!
+There are the features we are looking for:
```
Comment (smola) [inline content/post/siva.md]: :+1:

File: content/post/siva.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+  compressed.
+2. **Seekable.** We need random access to arbitrary positions of the archived
+   files.
+3. **Indexed.** Fast access to specific files.
+4. **Cheap Add.** With every update, we need to add two files
+  to the archive, so ideally, we should be able to add these files to the archive
+  without completely rewriting it.
+5. **Concatenable.** Ideally, it should be possible to add new files to the archive
+  with a single append operation. We could use [GCS compose](https://cloud.google.com/storage/docs/composite-objects)
+  or HDFS append.
```
Comment (smola) [inline content/post/siva.md]: Added clarification when explaining packfiles int he motivation, as well as in this paragraph.

File: content/post/siva.md (hunk 1 above)
Comment (smola) [inline content/post/siva.md]: :)

File: content/post/siva.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+Being tar the most widely used archiving format in UNIX-like systems, we decided
+to give it a chance.
+
+Let's have a look at the tar file structure, in a simplified way:
+
+![tar file structure](/post/siva/tar.png)
+
+A tar archive is essentially a sequence of files. Each file contains a header, padding
+(files are stored in blocks of 512 bytes) and file content. The end of file is
```
Comment (smola) [inline content/post/siva.md]: :+1:

File: content/post/siva.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+There is no index that can be used for random access to tar contents. This is
+natural since tar (*t*ape *ar*chive) was designed to write and read sequentially
+on tape drives. Given the ubiquity of the tar format, many people have added
+indexing to tar, either by using a separate file (e.g.
+[tarindexer](https://github.com/devsnd/tarindexer)) or embedding the index in
+the tar file itself (e.g. [rat](https://github.com/mcuadros/go-rat)). The later
+approach enables us to use the embedded index while preserving compatibility
+with standard tar tooling, since standard tar will ignore data after the EOF mark.
+The result would be as follows:
```
Comment (smola) [inline content/post/siva.md]: :+1:

File: content/post/siva.md [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+
+![tar file structure](/post/siva/tar_index.png)
+
+We can combine both indexing and concatenation. The result would be something
+like this:
+
+![tar file structure](/post/siva/tar_concat_index.png)
+
+Our conclusion is that we could use the tar format with some tricks to fulfill
+all our needs, and still be compatible with GNU tar or FreeBSD tar. It would
```
Comment (smola) [inline content/post/siva.md]: :+1:

File: content/post/siva.md [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+have the following drawbacks:
+
+1. Adding indexing and concatenation support would mean writing a custom tar
+   implementation is needed for every language that we use.
```
Comment (smola) [inline content/post/siva.md]: :+1:

File: content/post/siva.md [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+2. It is not space efficient: metadata is duplicated both in the index and headers,
```
Comment (smola) [inline content/post/siva.md]: :+1:

File: content/post/siva.md [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+And then there is a [myriad of more obscure archiving formats](https://en.wikipedia.org/wiki/List_of_archive_formats)
+(e.g. [mar](https://wiki.mozilla.org/Software_Update:MAR#Mozilla_ARchive),
+[zpaq](https://github.com/zpaq/zpaq)), most of which lack adequate documentation
+or multiple implementations. We decided that doing the full research to find an
+adequate format was not going to pay off, since we would need to write our own
+implementations in multiple languages anyway.
+
+## Enter śiva
+
+Here it is! The format that has all the features we wanted: [śiva](https://github.com/src-d/go-siva)
```
Comment (smola) [inline content/post/siva.md]: :+1:

File: content/post/siva.md (hunk 2 above)
Comment (smola) [inline content/post/siva.md]: :+1:

File: content/post/siva.md (hunk 5 above)
Comment (alcortesm) [inline content/post/siva.md]: :+1:

File: content/post/siva.md (hunk 7 above)
Comment (alcortesm) [inline content/post/siva.md]: :+1:

**Pull request merged**